### ISSUE TITLE
pcmanfm-qt don't start as root on live isos

### ISSUE THREAD

**Issue by agaida** [lxqt/pcmanfm-qt#958]
it don't start - without a meaningful message

##### Expected Behavior

it should start as root

##### Current Behavior

it don't

##### Possible Solution

fix it

##### Steps to Reproduce (for bugs)

1. get a live iso (siduction, manjaro, debian buster) 
2.  start and open a terminal
3. lx(sudo) pcmanfm-qt
4. start using swearwords

##### Context

I want ot use pcmanfm-qt as root even on live isos - not possible
(thanks adrian from rescatux for spotting this)

##### System Information
latest release

**Comment by agaida** [comment]
btw - not funny, i tried to argue that it should work and the bug sit in front of the screen - i was wrong :sunglasses:

**Comment by agaida** [comment]
* http://rwthaachen.dl.osdn.jp/storage/g/m/ma/manjaro-community/lxqt/18.0.4-19104/manjaro-lxqt-18.0.4-19104-stable-x86_64.iso
* https://isobuilds.siduction.org/lxqt/amd64/siduction-18.3.0-patience-lxqt-amd64-201905071610.iso
* https://cdimage.debian.org/cdimage/unofficial/non-free/cd-including-firmware/weekly-live-builds/amd64/iso-hybrid/debian-live-testing-amd64-lxqt+nonfree.iso

**Comment by agaida** [comment]
btw - only a heads up - lxsudo lxqt-about works fine, lxsudo lximage-qt don't too in manjaro

**Comment by tsujan** [comment]
`lxsudo dbus-launch %s`

If that works, this is a duplicate of https://github.com/lxqt/pcmanfm-qt/issues/774 and can be closed.

If not, tell me.

**Comment by agaida** [comment]
ok, works - but why does it work in a real installation but not in a live iso

**Comment by tsujan** [comment]
As I said in https://github.com/lxqt/pcmanfm-qt/issues/774, I don't know the reason but I encountered it years before pcmanfm-qt was created or the idea of LXQt existed ;)

What happens in Debian Stable has already happened in Unstable (my previous distro) and what shows up in Debian Unstable is already known to Arch users. Lots of people know about `dbus-launch` but I added a short explanation to pcmanfm-qt.

Not our problem.

**Comment by agaida** [comment]
btw - and we might even use the dbus-start in our own programming - starting a pcmanfm-qt root session from within pcmanfm-qt is also impossible in a live iso - but here a user don't have any chances to work around. So the ball is in our side of the field.

**Comment by agaida** [comment]
and the explanation is to late for the upcoming stable - ok, i might port back, but we should at least make sure that opening the root instance from within pcmanfm-qt work correct anywhere.

**Comment by tsujan** [comment]
We can't do anything more than this (which is done):

![dbus](https://user-images.githubusercontent.com/981076/57561407-be872400-73a0-11e9-95a4-65035146d501.png)

**Comment by agaida** [comment]
ok - to have dbus-launch, dbus-x11 must be installed in debian

**Comment by tsujan** [comment]
Yes, that's also the case on Manjaro (or was) but a simple `dbus-launch` is not a big deal, especially on systems that have a Wayland session too.

**Comment by agaida** [comment]
it is a big deal when debian is in deep freeze :D

**Comment by agaida** [comment]
it means:
* have a bug for it
* fix the bug with a minimal patch
* upload to unstable
* file a bug for the release team
* hope that the release team ack the changes made and unblock the fixed package for buster

**Comment by tsujan** [comment]
If I were you, I'd do nothing. No one thought it was a bug when it first happened years ago.

**Comment by tsujan** [comment]
IMO, you're too hard on yourself sometimes. This isn't a big deal at all.

**Comment by agaida** [comment]
it remind me of my very beginnings with linux in 2009 - and yes, now i read it agian and agian - i remember the infamous dbus thing. Maybe i'm a bit to picky sometimes, but i didn't buy the "it was hard for us so it is ok when it is hard for newbies" - and ok, i was trapped with dbus-launch too. I will patch the dbus-launch thing into pcmanfm-qt and be done with.

**Comment by tsujan** [comment]
Since "Wayland will come soon" (= not soon), everyone has to be ready. dbus hasn't been an exception.

I think, when I used Debian, dbus and dbus-x11 were installed intermittently because of some dependencies (but Debian didn't have dbus-x11 then) or, maybe, developers changed their decision intermittently; I don't know. But the fact is that wherever you see X11, it's supposed to be "temporary".

> it was hard for us so it is ok when it is hard for newbies

It wasn't hard for us. The first time I encountered it, a simple googling solved the problem in a few minutes. Newbies know what Google is ;)

**Comment by tsujan** [comment]
Oh, and please test `dbus-launch` with dbus-x11 too. I've never done so but I think it should work.

**Comment by agaida** [comment]
```
% apt-file search bin/dbus-launch 
dbus-1-dbg: /usr/lib/x86_64-linux-gnu/dbus-1.0/debug-build/bin/dbus-launch
dbus-tests: /usr/lib/dbus-1.0/debug-build/bin/dbus-launch
dbus-x11: /usr/bin/dbus-launch
```
So we need this package (dbus-x11) installed. And i was thinking of a test in pcmanfm-qt - it is dead easy. One need a working dbus address to start pcmanfm-qt. My idea was simple:

```
if -z "$DBUS_SESSION_BUS_ADDRESS"  --> dbus-launch and be done with
```

**Comment by tsujan** [comment]
> So we need this package (dbus-x11) installed.

I don't think that's a good idea. The user might need to replace it with dbus.

I also think we shouldn't do anything in our apps in this regard for the same reason.

**Comment by agaida** [comment]
it was only a idea :D

**Comment by tsujan** [comment]
My point, from start: It's not our fault; it's not your fault; it's not anyone's fault :)

**Comment by tsujan** [comment]
Déjà vu: https://github.com/lxqt/lxqt/issues/1224#issuecomment-271078967

**Comment by tsujan** [comment]
OK, I admit I misunderstood the whole Debian situation:

In Arch:
`dbus-launch` is provided by `dbus` and `lxsudo dbus-launch pcmanfm-qt` always works.
`lxsudo pcmanfm-qt` should work with `dbus-x11` (haven't tried it though) but `dbus-x11` has conflict with `dbus`.

In Debian:
`dbus-launch` is provided by `dbus-x11`, which depends on `dbus`.
`lxsudo pcmanfm-qt` never works but `lxsudo dbus-launch pcmanfm-qt` works *only when dbus-x11 is installed*.

Very confusing to me.

**Comment by agaida** [comment]
thats dead easy - try `echo $DBUS_SESSION_BUS_ADDRESS` - hey, we have a dbus session -try it with `sudo echo $DBUS_SESSION_BUS_ADDRESS` or become root and try the echo - you will no dbus address in Manjaro nor Debian in the live iso - no try dbus-launch - oha - it create a temporary dbus session or pick up a existing - ok, long story short - there are a few parameters :D

**Comment by tsujan** [comment]
> thats dead easy

Not to me. I don't know about ISOs but `dbus-x11` has completely different meanings in Debian and Arch. In Arch, it isn't needed for launching a root instance of pcmanfm-qt.

The situation becomes more complex under Wayland. As if running root instances has become a taboo. `admin:///` is good for nothing because, apparently, only apps supporting gvfs work with it (you can't open a root instance of FeatherPad with pcmanfm-qt+`admin:///`. I don't know about a Qt solution outside KDE either… but I hope a unified solution for all apps will be possible at some point.

**Comment by tsujan** [comment]
pcmanfm-qt's `admin:///` with Kate:

![admin](https://user-images.githubusercontent.com/981076/57564337-7aefe280-73bf-11e9-9dee-cd1b745eea1b.png)

**Comment by agaida** [comment]
@tsujan ok - we should remove the dbus-launch thing, there are reasons why it is in the dbus-x11 package in debian. Instead we should evaluate the usage of dbus-run-session.

Steps to reproduce:
* Run the buster image
* open a terminal
* sudo dbus-run-session -- pcmanfm-qt

No additional steps needed, all needed dependencies are in the debian and manjaro - so i guess we should use the "modern" way to do this.

**Comment by tsujan** [comment]
> we should remove the dbus-launch thing

`dbus-launch` is relevant on Arch and its derivatives. If you suggest an extra GUI explanation (without removing the current one), tell me to add it.

BTW, having got disappointed by various artificial obstacles to using root instances, I tried "the superior solution" with FeatherPad and Kate, i.e. `sudoedit`. In short, it's crap, at least in 2 ways.

**Comment by agaida** [comment]
erm, no, it isn't that relevant - seriously

![](https://i.imgur.com/bLgtUMN.png)

**Comment by agaida** [comment]
RE: sudoedit - it was recommended by some people who write code as `if pid=0 then exit;` - what do you expect?

**Comment by tsujan** [comment]
Yes, `lxsudo dbus-run-session -- %s` also works in pcmanfm-qt (I think `sudo...` should never be used):

![dbus-run-session](https://user-images.githubusercontent.com/981076/57572141-b9bb8200-742b-11e9-9923-4d96b1d50bc8.png)

**Comment by tsujan** [comment]
We can change the explanation to:
```
For switching user: lxsudo dbus-run-session -- %s (or, probably, lxsudo dbus-launch %s or lxsudo %s)
```
And set it to the first by default. ?

**Comment by agaida** [comment]
if i remember correctly `dbus-lauch` was splitted out because of X11 dependencies - we should not silently re-introduce them if not needed.  And yes `sudo` is ten years successfully trained bad behaviour :sunglasses:

**Comment by tsujan** [comment]
So, your suggestion would be `lxsudo dbus-run-session -- %s` and nothing else in the explanation? Should we proceed with that (I won't disagree if you say "yes" — we need a solution that works everywhere).

**Comment by agaida** [comment]
i will do so in debian - because we have a permanently running system dbus and sessions based on the user session. Looks like
```
agaida@ramme ~ % ps faux | grep dbus
message+  1144  0.0  0.0  10644  5968 ?        Ss   17:54   0:00 /usr/bin/dbus-daemon --system --address=systemd: --nofork --nopidfile --systemd-activation --syslog-only
agaida   17612  0.0  0.0   9940  5312 ?        Ss   17:54   0:00  \_ /usr/bin/dbus-daemon --session --address=systemd: --nofork --nopidfile --systemd-activation --syslog-only
agaida   17798  0.0  0.0   9296  4504 ?        S    17:54   0:00  |   \_ /usr/bin/dbus-daemon --config-file=/usr/share/defaults/at-spi2/accessibility.conf --nofork --print-address 3
agaida   17970  0.0  0.1 372152 62988 ?        Sl   17:54   0:00  \_ /usr/bin/python3 /usr/share/system-config-printer/scp-dbus-service.py
agaida    4576  0.3  0.1 325100 49460 pts/0    Sl+  18:08   0:00      \_ lxsudo dbus-run-session -- pcmanfm-qt
root      4579  0.0  0.0   7404  4024 pts/1    Ss+  18:08   0:00          \_ sudo -E /bin/sh -c unset LC_ALL; exec 'dbus-run-session' '--' 'pcmanfm-qt'
root      4583  0.0  0.0   5004  1172 pts/1    S+   18:08   0:00              \_ dbus-run-session -- pcmanfm-qt
root      4584  0.0  0.0   9184  4524 pts/1    S+   18:08   0:00                  \_ dbus-daemon --nofork --print-address 4 --session
```
I guess i would do so - maybe double check with fedora and suse - but yes, i would.

**Comment by agaida** [comment]
The best is - i prepred a PR in that direction.

**Comment by tsujan** [comment]
```
[pedram@manjaro:~]$ ps faux | grep dbus
dbus       747  0.0  0.0  12144  5484 ?        Ss   May09   1:40 /usr/bin/dbus-daemon --system --address=systemd: --nofork --nopidfile --systemd-activation --syslog-only
pedram   19682  0.0  0.0  11576  5024 ?        Ss   May10   0:43  \_ /usr/bin/dbus-daemon --session --address=systemd: --nofork --nopidfile --systemd-activation --syslog-only
pedram   19987  0.0  0.0  10556  3920 ?        S    May10   0:09  |   \_ /usr/bin/dbus-daemon --config-file=/usr/share/defaults/at-spi2/accessibility.conf --nofork --print-address 3
pedram   19902  0.0  0.0   6144  2276 pts/0    S+   20:42   0:00      \_ grep dbus
```

> The best is - i prepred a PR in that direction.

Do it! What about systems like Gentoo? (I ask because I don't know.)

**Comment by tsujan** [comment]
And please add something short like "sudo alone should never be used". People use it and the permissions of their user settings are changed.

**Comment by agaida** [comment]
ok, gentoo is a special case, void linux will be too, devuan too - to be honest, i don't care much. i just hope that the standard ways to start a dbus session will work, see our first line with --systemd-activation. Maybe more precise: I don't care which initial mechanism is used to start the root dbus-daemon --system.

**Comment by tsujan** [comment]
OK. Gentoo users will find their way. The privilege command can be edited after all.

**Comment by tsujan** [comment]
There's another good thing about `dbus-run-session` too: previously, when I closed a root instance, root gvfs process were persistent until logging out; now, they're stopped after closing the root instance.
